Entra ID support for auto config

TheovanKraay opened the pull request:
Add support for Entra ID in the auto config for Azure Cosmos DB vector store support. 
This PR also adds a few fixes/enhancements and corresponding tests:

- fix to ensure getMetadata() returns the metadata fields and values
- allow metadata field to be defined as the Cosmos DB partition key (and not just /id)
- allow choice between gateway and direct mode in autoconfig

ilayaperumalg commented:
@TheovanKraay Thanks for the PR!  Could you complete the DCO check (by signing off your commit: git commit --amend --signoff) and fix the PR build failure (there is a format issue)?

TheovanKraay commented:
Thank @ilayaperumalg I've done as you requested. Let me know if there are other issues blocking merge/release?

TheovanKraay commented:
Hi @ilayaperumalg and all - any eta on when this could be merged and released? I have some other fixes/enhancements I want to make behind this one, but doing them in small increments. Thanks!

TheovanKraay commented:
@ilayaperumalg since the PR is not merged yet, I added another small fix to expose metadata fields in response. I have some more changes planned but will hold off now until this is merged. Hoping it can be pushed soon. Thanks again.

markpollack commented:
@TheovanKraay we have been without access to cosmos db for quite some time.  The current code on the main branch is failing the test `CosmosDBVectorStoreIT`  error is below.


Could you please confirm that the current main branch has or doesn't have this issue.  I am not sure how to debug this.



```
{
  "ClassName": "CosmosException",
  "userAgent": "azsdk-java-cosmos/4.63.2 Linux/6.11.0-21-generic JRE/21.0.5",
  "statusCode": 400,
  "resourceAddress": "https://spring-ai-comosdb1-westus3.documents.azure.com:443/dbs/test-database/colls",
  "error": "{\"code\":\"BadRequest\",\"message\":\"Message: {\\\"Errors\\\":[\\\"One of the specified inputs is invalid\\\"]}\\r\\nActivityId: 2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Request URI: /apps/a277a4f2-3b35-4a1f-bd72-a2d727d0d941/services/0feba6ad-ee3e-476e-9c16-3fb20f103afe/partitions/0ca0768d-9474-44ab-a374-f16a0e663274/replicas/133883487107516521p, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.14.0, StatusCode: BadRequest\",\"additionalErrorInfo\":null}",
  "innerErrorMessage": "Message: {\"Errors\":[\"One of the specified inputs is invalid\"]}\r\nActivityId: 2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, , SDK: Microsoft.Azure.Documents.Common/2.14.0, StatusCode: BadRequest",Request URI: /apps/a277a4f2-3b35-4a1f-bd72-a2d727d0d941/services/0feba6ad-ee3e-476e-9c16-3fb20f103afe/partitions/0ca0768d-9474-44ab-a374-f16a0e663274/replicas/133883487107516521p, RequestStats: 
  "causeInfo": null,
  "responseHeaders": "{Transfer-Encoding=chunked, x-ms-last-state-change-utc=Mon, 07 Apr 2025 01:22:50.163 GMT, x-ms-current-replica-set-size=4, Server=Microsoft-HTTPAPI/2.0, x-ms-request-duration-ms=0.516, x-ms-session-token=0:-1#9, lsn=9, x-ms-request-charge=0, x-ms-schemaversion=1.19, x-ms-transport-request-id=179465, x-ms-number-of-read-regions=0, x-ms-current-write-quorum=3, x-ms-cosmos-quorum-acked-llsn=9, x-ms-quorum-acked-lsn=9, Date=Wed, 09 Apr 2025 20:05:47 GMT, x-ms-activity-id=2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Strict-Transport-Security=max-age=31536000, x-ms-xp-role=1, x-ms-global-Committed-lsn=9, x-ms-cosmos-llsn=9, x-ms-gatewayversion=version=2.14.0, x-ms-serviceversion=version=2.14.0.0, Content-Type=application/json}",
  "requestHeaders": "[x-ms-cosmos-sdk-supportedcapabilities=1, Accept=application/json, x-ms-date=Wed, 09 Apr 2025 20:05:48 GMT, x-ms-offer-throughput=1000, Content-Type=application/json]",
  "cosmosDiagnostics": {
    "userAgent": "azsdk-java-cosmos/4.63.2 Linux/6.11.0-21-generic JRE/21.0.5 SpringAI-CDBNoSQL-VectorStore",
    "activityId": "2f3ccc4a-b37e-4f34-82ce-bce11caa7b92",
    "requestLatencyInMs": 94,
    "requestStartTimeUTC": "2025-04-09T20:05:48.412595553Z",
    "requestEndTimeUTC": "2025-04-09T20:05:48.506983682Z",
    "responseStatisticsList": [],
    "supplementalResponseStatisticsList": [],
    "addressResolutionStatistics": {},
    "regionsContacted": [
      "west us 3"
    ],
    "retryContext": {
      "statusAndSubStatusCodes": null,
      "retryLatency": 0,
      "retryCount": 0
    },
    "metadataDiagnosticsContext": {
      "metadataDiagnosticList": null
    },
    "serializationDiagnosticsContext": {
      "serializationDiagnosticsList": [
        {
          "serializationType": "CONTAINER_SERIALIZATION",
          "startTimeUTC": "2025-04-09T20:05:48.413421547Z",
          "endTimeUTC": "2025-04-09T20:05:48.413983068Z",
          "durationInMilliSecs": 0.561521
        }
      ]
    },
    "gatewayStatisticsList": [
      {
        "sessionToken": "0:-1#9",
        "operationType": "Create",
        "resourceType": "DocumentCollection",
        "statusCode": 400,
        "subStatusCode": 0,
        "requestCharge": 0,
        "requestTimeline": [
          {
            "eventName": "connectionAcquired",
            "startTimeUTC": "2025-04-09T20:05:48.415137142Z",
            "durationInMilliSecs": 0.601975
          },
          {
            "eventName": "connectionConfigured",
            "startTimeUTC": "2025-04-09T20:05:48.415739117Z",
            "durationInMilliSecs": 0.103313
          },
          {
            "eventName": "requestSent",
            "startTimeUTC": "2025-04-09T20:05:48.415842430Z",
            "durationInMilliSecs": 1.310173
          },
          {
            "eventName": "transitTime",
            "startTimeUTC": "2025-04-09T20:05:48.417152603Z",
            "durationInMilliSecs": 88.328356
          },
          {
            "eventName": "received",
            "startTimeUTC": "2025-04-09T20:05:48.505480959Z",
            "durationInMilliSecs": 0.666107
          }
        ],
        "partitionKeyRangeId": null,
        "responsePayloadSizeInBytes": 0,
        "exceptionMessage": "Message: {\"Errors\":[\"One of the specified inputs is invalid\"]}\r\nActivityId: 2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Request URI: /apps/a277a4f2-3b35-4a1f-bd72-a2d727d0d941/services/0feba6ad-ee3e-476e-9c16-3fb20f103afe/partitions/0ca0768d-9474-44ab-a374-f16a0e663274/replicas/133883487107516521p, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.14.0, StatusCode: BadRequest",
        "exceptionResponseHeaders": "{Transfer-Encoding=chunked, x-ms-last-state-change-utc=Mon, 07 Apr 2025 01:22:50.163 GMT, x-ms-current-replica-set-size=4, Server=Microsoft-HTTPAPI/2.0, x-ms-request-duration-ms=0.516, x-ms-session-token=0:-1#9, lsn=9, x-ms-request-charge=0, x-ms-schemaversion=1.19, x-ms-transport-request-id=179465, x-ms-number-of-read-regions=0, x-ms-current-write-quorum=3, x-ms-cosmos-quorum-acked-llsn=9, x-ms-quorum-acked-lsn=9, Date=Wed, 09 Apr 2025 20:05:47 GMT, x-ms-activity-id=2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Strict-Transport-Security=max-age=31536000, x-ms-xp-role=1, x-ms-global-Committed-lsn=9, x-ms-cosmos-llsn=9, x-ms-gatewayversion=version=2.14.0, x-ms-serviceversion=version=2.14.0.0, Content-Type=application/json}",
        "locationToLocationSpecificHealthContext": {
          "v": null
        }
      }
    ],
    "samplingRateSnapshot": 1,
    "bloomFilterInsertionCountSnapshot": 0,
    "systemInformation": {
      "usedMemory": "168823 KB",
      "availableMemory": "24472713 KB",
      "systemCpuLoad": "empty",
      "availableProcessors": 32
    },
    "clientCfgs": {
      "id": 1,
      "machineId": "uuid:7ec0af51-02c7-4490-bb72-9917087e3df6",
      "connectionMode": "GATEWAY",
      "numberOfClients": 0,
      "excrgns": "[]",
      "clientEndpoints": {
        "https://spring-ai-comosdb1.documents.azure.com:443/": 1
      },
      "connCfg": {
        "rntbd": null,
        "gw": "(cps:1000, nrto:PT1M, icto:PT1M, p:false)",
        "other": "(ed: true, cs: false, rv: true)"
      },
      "consistencyCfg": "(consistency: null, mm: true, prgns: [])",
      "proactiveInitCfg": "",
      "e2ePolicyCfg": "",
      "sessionRetryCfg": "",
      "partitionLevelCircuitBreakerCfg": "(cb: false, type: CONSECUTIVE_EXCEPTION_COUNT_BASED, rexcntt: 10, wexcntt: 5)"
    }
  }
}
```

TheovanKraay commented:
> @TheovanKraay we have been without access to cosmos db for quite some time. The current code on the main branch is failing the test `CosmosDBVectorStoreIT` error is below.
> 
> Could you please confirm that the current main branch has or doesn't have this issue. I am not sure how to debug this.
> 
> ```
> {
>   "ClassName": "CosmosException",
>   "userAgent": "azsdk-java-cosmos/4.63.2 Linux/6.11.0-21-generic JRE/21.0.5",
>   "statusCode": 400,
>   "resourceAddress": "https://spring-ai-comosdb1-westus3.documents.azure.com:443/dbs/test-database/colls",
>   "error": "{\"code\":\"BadRequest\",\"message\":\"Message: {\\\"Errors\\\":[\\\"One of the specified inputs is invalid\\\"]}\\r\\nActivityId: 2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Request URI: /apps/a277a4f2-3b35-4a1f-bd72-a2d727d0d941/services/0feba6ad-ee3e-476e-9c16-3fb20f103afe/partitions/0ca0768d-9474-44ab-a374-f16a0e663274/replicas/133883487107516521p, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.14.0, StatusCode: BadRequest\",\"additionalErrorInfo\":null}",
>   "innerErrorMessage": "Message: {\"Errors\":[\"One of the specified inputs is invalid\"]}\r\nActivityId: 2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, , SDK: Microsoft.Azure.Documents.Common/2.14.0, StatusCode: BadRequest",Request URI: /apps/a277a4f2-3b35-4a1f-bd72-a2d727d0d941/services/0feba6ad-ee3e-476e-9c16-3fb20f103afe/partitions/0ca0768d-9474-44ab-a374-f16a0e663274/replicas/133883487107516521p, RequestStats: 
>   "causeInfo": null,
>   "responseHeaders": "{Transfer-Encoding=chunked, x-ms-last-state-change-utc=Mon, 07 Apr 2025 01:22:50.163 GMT, x-ms-current-replica-set-size=4, Server=Microsoft-HTTPAPI/2.0, x-ms-request-duration-ms=0.516, x-ms-session-token=0:-1#9, lsn=9, x-ms-request-charge=0, x-ms-schemaversion=1.19, x-ms-transport-request-id=179465, x-ms-number-of-read-regions=0, x-ms-current-write-quorum=3, x-ms-cosmos-quorum-acked-llsn=9, x-ms-quorum-acked-lsn=9, Date=Wed, 09 Apr 2025 20:05:47 GMT, x-ms-activity-id=2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Strict-Transport-Security=max-age=31536000, x-ms-xp-role=1, x-ms-global-Committed-lsn=9, x-ms-cosmos-llsn=9, x-ms-gatewayversion=version=2.14.0, x-ms-serviceversion=version=2.14.0.0, Content-Type=application/json}",
>   "requestHeaders": "[x-ms-cosmos-sdk-supportedcapabilities=1, Accept=application/json, x-ms-date=Wed, 09 Apr 2025 20:05:48 GMT, x-ms-offer-throughput=1000, Content-Type=application/json]",
>   "cosmosDiagnostics": {
>     "userAgent": "azsdk-java-cosmos/4.63.2 Linux/6.11.0-21-generic JRE/21.0.5 SpringAI-CDBNoSQL-VectorStore",
>     "activityId": "2f3ccc4a-b37e-4f34-82ce-bce11caa7b92",
>     "requestLatencyInMs": 94,
>     "requestStartTimeUTC": "2025-04-09T20:05:48.412595553Z",
>     "requestEndTimeUTC": "2025-04-09T20:05:48.506983682Z",
>     "responseStatisticsList": [],
>     "supplementalResponseStatisticsList": [],
>     "addressResolutionStatistics": {},
>     "regionsContacted": [
>       "west us 3"
>     ],
>     "retryContext": {
>       "statusAndSubStatusCodes": null,
>       "retryLatency": 0,
>       "retryCount": 0
>     },
>     "metadataDiagnosticsContext": {
>       "metadataDiagnosticList": null
>     },
>     "serializationDiagnosticsContext": {
>       "serializationDiagnosticsList": [
>         {
>           "serializationType": "CONTAINER_SERIALIZATION",
>           "startTimeUTC": "2025-04-09T20:05:48.413421547Z",
>           "endTimeUTC": "2025-04-09T20:05:48.413983068Z",
>           "durationInMilliSecs": 0.561521
>         }
>       ]
>     },
>     "gatewayStatisticsList": [
>       {
>         "sessionToken": "0:-1#9",
>         "operationType": "Create",
>         "resourceType": "DocumentCollection",
>         "statusCode": 400,
>         "subStatusCode": 0,
>         "requestCharge": 0,
>         "requestTimeline": [
>           {
>             "eventName": "connectionAcquired",
>             "startTimeUTC": "2025-04-09T20:05:48.415137142Z",
>             "durationInMilliSecs": 0.601975
>           },
>           {
>             "eventName": "connectionConfigured",
>             "startTimeUTC": "2025-04-09T20:05:48.415739117Z",
>             "durationInMilliSecs": 0.103313
>           },
>           {
>             "eventName": "requestSent",
>             "startTimeUTC": "2025-04-09T20:05:48.415842430Z",
>             "durationInMilliSecs": 1.310173
>           },
>           {
>             "eventName": "transitTime",
>             "startTimeUTC": "2025-04-09T20:05:48.417152603Z",
>             "durationInMilliSecs": 88.328356
>           },
>           {
>             "eventName": "received",
>             "startTimeUTC": "2025-04-09T20:05:48.505480959Z",
>             "durationInMilliSecs": 0.666107
>           }
>         ],
>         "partitionKeyRangeId": null,
>         "responsePayloadSizeInBytes": 0,
>         "exceptionMessage": "Message: {\"Errors\":[\"One of the specified inputs is invalid\"]}\r\nActivityId: 2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Request URI: /apps/a277a4f2-3b35-4a1f-bd72-a2d727d0d941/services/0feba6ad-ee3e-476e-9c16-3fb20f103afe/partitions/0ca0768d-9474-44ab-a374-f16a0e663274/replicas/133883487107516521p, RequestStats: , SDK: Microsoft.Azure.Documents.Common/2.14.0, StatusCode: BadRequest",
>         "exceptionResponseHeaders": "{Transfer-Encoding=chunked, x-ms-last-state-change-utc=Mon, 07 Apr 2025 01:22:50.163 GMT, x-ms-current-replica-set-size=4, Server=Microsoft-HTTPAPI/2.0, x-ms-request-duration-ms=0.516, x-ms-session-token=0:-1#9, lsn=9, x-ms-request-charge=0, x-ms-schemaversion=1.19, x-ms-transport-request-id=179465, x-ms-number-of-read-regions=0, x-ms-current-write-quorum=3, x-ms-cosmos-quorum-acked-llsn=9, x-ms-quorum-acked-lsn=9, Date=Wed, 09 Apr 2025 20:05:47 GMT, x-ms-activity-id=2f3ccc4a-b37e-4f34-82ce-bce11caa7b92, Strict-Transport-Security=max-age=31536000, x-ms-xp-role=1, x-ms-global-Committed-lsn=9, x-ms-cosmos-llsn=9, x-ms-gatewayversion=version=2.14.0, x-ms-serviceversion=version=2.14.0.0, Content-Type=application/json}",
>         "locationToLocationSpecificHealthContext": {
>           "v": null
>         }
>       }
>     ],
>     "samplingRateSnapshot": 1,
>     "bloomFilterInsertionCountSnapshot": 0,
>     "systemInformation": {
>       "usedMemory": "168823 KB",
>       "availableMemory": "24472713 KB",
>       "systemCpuLoad": "empty",
>       "availableProcessors": 32
>     },
>     "clientCfgs": {
>       "id": 1,
>       "machineId": "uuid:7ec0af51-02c7-4490-bb72-9917087e3df6",
>       "connectionMode": "GATEWAY",
>       "numberOfClients": 0,
>       "excrgns": "[]",
>       "clientEndpoints": {
>         "https://spring-ai-comosdb1.documents.azure.com:443/": 1
>       },
>       "connCfg": {
>         "rntbd": null,
>         "gw": "(cps:1000, nrto:PT1M, icto:PT1M, p:false)",
>         "other": "(ed: true, cs: false, rv: true)"
>       },
>       "consistencyCfg": "(consistency: null, mm: true, prgns: [])",
>       "proactiveInitCfg": "",
>       "e2ePolicyCfg": "",
>       "sessionRetryCfg": "",
>       "partitionLevelCircuitBreakerCfg": "(cb: false, type: CONSECUTIVE_EXCEPTION_COUNT_BASED, rexcntt: 10, wexcntt: 5)"
>     }
>   }
> }
> ```

Its failing in main because this code:

```java
if (this.partitionKeyPath == null) {
	partitionKeyPath = "/id";
}
```

...updates the local `partitionKeyPath` variable instead of the instance variable of the same name. In the [original version](https://github.com/spring-projects/spring-ai/blob/7b06fcf98b2daf34f72870e058cb1030eb91717d/vector-stores/spring-ai-azure-cosmos-db-store/src/main/java/org/springframework/ai/vectorstore/CosmosDBVectorStore.java) of the code there was no instance variable, but I see this has been refactored a lot since I added it. 

The test fails because it doesn't set `partitionKeyPath`, and since there is nothing to properly default to id if it's not set, that results in the error (users would not get the error as long as they set `partitionKeyPath` when creating the VectorStore bean, though currently they would be restricted to setting it to "/id" only). This is one of the bugs I've fixed in this PR, along with enhancing it so that a `partitionKeyPath`  other than "/id" can be set - as long as it is within "/metadata" - which will have to be the case for now - separate conversation around this use case refers (tx plus vectors together).

TheovanKraay commented:
@ilayaperumalg @markpollack I messed up this PR so created a new one: https://github.com/spring-projects/spring-ai/pull/2707

sobychacko commented:
@TheovanKraay Thanks for the new PR - we will continue the discussion there.
